### PR TITLE
RC2 Updates

### LINKED ISSUE #97: RC2 Updates
Update repo to fix issues present in the RC2 release these include:

- Update to daffodil 3.3.0
- Update build/bin.NOTICE
- Update dependencies versions, as long as they don't break anything.

### PR DESCRIPTION
**Pull request by Shanedell**
Multiple updates from rc2 release:

- Update daffodilVersion from 3.2.1 to 3.3.0.
- Add plugins to project/plugins.sbt.
  - These plugins allow for checking dependencies for updates, CVEs and viewing the tree of dependencies.
- Update scalaVersion from 2.12.13 to 2.12.15.
  - Can move to 2.13.8 at some point not sure when we want to do this however.
- Update "logback-classic" version from 1.2.3 to 1.2.11 to fix CVEs.
- Update "com.microsoft.java.debug.core" version from 0.31.1 to 0.35.0.
- Update "fs2-io" version from 3.0.4 to 3.0.6.
  - Updating to 3.2.5 breaks the build.
- Update "decline-effect" from 2.1.0 to 2.2.0.
- Update "log4cats-slf4j" from 2.1.0 to 2.1.1.
  - Updating to 2.2.0 breaks the build.
  - CVEs seem to be false positives here.
- Add dependencyOverrides for commons-lang3 to fix a CVE item.
- Update bin.NOTICE files as some of the incorated projects NOTICE's were updated.

Closes #97

### CODE REVIEW

**Review by tuxji — APPROVED** [review]
+1

**Review by mbeckerle — APPROVED** [review]
+1

File: build/bin.NOTICE
Comment (stevedlawrence) [inline build/bin.NOTICE:187]: We just made a bunch of corrections to this file that are accurate for 3.3.0. You might need to incorporate some of those changes.

File: project/plugins.sbt
Comment (stevedlawrence) [inline project/plugins.sbt:37]: Personally, I think these plugins should be part of ~/.sbt/plugins/plugins.sbt rather than the repo since these aren't strictly needed for the vscode repo. Minimizes dependencies for people that don't care about this this stuff. But I don't feel that strongly about it.

**Review by stevedlawrence — COMMENTED** [review]
+1, might need to pull in changes or 3.3.0

**Comment by stevedlawrence** [comment]
Also, In general we try to ensure every PR has an associated Issue associated with it for tracking purposes. It would be good to create an issue for these changes and and add a `Closes` line to the commit message.

File: build/bin.NOTICE
Comment (Shanedell) [reply build/bin.NOTICE:187]: @stevedlawrence So I updated this on the weekend to contain the contents I found in `daffodil/NOTICE` was that not the correct fix for this?

Comment (stevedlawrence) [reply build/bin.NOTICE:187]: That was correct, but @tuxji opened a new PR (https://github.com/apache/daffodil/pull/777) to fix some NOTICE issues we found during the Daffodil 3.3.0 vote. We decided it wasn't worth blocking the release and fixed them shortly after. That PR hasn't been merged yet, but should probably be merged relatively soon. At which point the relevant parts can be copied in. Note that PR also makes a minor update to the license file too.

Comment (Shanedell) [reply build/bin.NOTICE:187]: So I am confused. That PR is for updating the `bin.NOTICE` and `bin.LICENSE` files for Daffodil. These wouldn't be inside of the `bin.NOTICE` and `bin.LICENSE` correct? Just the root `NOTICE` and `LICENSE` file contents should be inside of the `bin.` files inside of `daffodil-vscode` correct? As the `bin` files between the two should differ as the daffodil-debugger Scala code doesn't use the same dependencies as the daffodil project.

Comment (Shanedell) [reply build/bin.NOTICE:187]: Okay I can work on getting these changes updated, should I just the contents of that PR or whats in main?

Comment (stevedlawrence) [reply build/bin.NOTICE:187]: You might wait until things are merged. I'm not sure what the final PR will look like, and the more similar the two projects are, the easier it is to verify.

Comment (Shanedell) [reply build/bin.NOTICE:187]: Okay that makes sense, do you want those other dependencies removed or anything else updated before I merge in this PR in? After merge would I be good to release RC3 or do you think I should wait till that daffodil PR is merged than fix the bin files and make another PR or keep this one open till that time?

Comment (stevedlawrence) [reply build/bin.NOTICE:187]: I think the dependency changes look fine. RC3 should be good once this is merged, but I might make sense to include the bin.NOTICE and bin.LICENSE updates in this PR. Also, did you confirm that the updated dependencies did not have a change to their LICENSE or NOTICE files that needs to be incorporated into bin.NOTICE/bin.LICENE?

Comment (Shanedell) [reply build/bin.NOTICE:187]: Yeah for all the ones I could fine most didn't change or if they did it was normally a simple bump of copyright year. The main thing was not knowing what I needed to copy over from daffodil files.

Comment (Shanedell) [reply build/bin.NOTICE:187]: @stevedlawrence Hey Steve, I am going through and updating stuff here so if something doesn't have a NOTICE but has a LICENSE it can be excluded from `bin.NOTICE` if the license is added to `bin.LICENSE` correct? What about something that has neither LICENSE or NOTICE does that just not get added to either file? What if the LICENSE is an Apache license does that need added to that file or we bypass those?

Comment (Shanedell) [reply build/bin.NOTICE:187]: So we have these items included but they aren't in either `bin.NOTICE` or `bin.LICENSE` but have Apache Licenses
```
org.typelevel.cats-effect
org.typelevel.cats-effect-kernel
org.typelevel.cats-effect-std
com.monovore.decline-effect
com.monovore.decline
org.typelevel.log4cats-core
org.typelevel.log4cats-slf4j
com.google.code.gson.gson
org.typelevel.literally
```
This one is in neither and has neither
```
org.typelevel.simulacrum-scalafix-annotations
```
So what do you think is best for these ones just leave them out? Adding the Apache license for all of the above, some are in the same project so it lowers the number down a bit, will cause that files to a bit more lengthy

Comment (stevedlawrence) [reply build/bin.NOTICE:187]: For dependencies that are under the Apache License, you must add the dependencies NOTICE content to bin.NOTICE. If it does not have a NOTICE file then you do not need to do anything for that dependency. And because it's ALv2 license, you not need to add anything to the bin.LICENSE file since it's the same as the vscode license.

For dependencies that are under a license other than the Apache License, you must add its LICENSE information to the bin.LICENSE file. Nothing should be added to the bin.NOTICE file, unless the license is a [Category B](https://www.apache.org/legal/resolved.html#category-b) license, but we don't see those very often, and in most cases the information should be in a README rather than the bin.NOTICE.

If a dependency does not have a LICENSE file, then we need to figure out what its license is and handle it according to the above rules. For example, sometimes the license is just mentioned in a README or file headers. If we cannot determine the license, it is assumed to be [Category X](https://www.apache.org/legal/resolved.html#category-x) and we cannot use the dependency until the author/authors specify a license and we determine its compatibility with ASF licensing.

Comment (stevedlawrence) [reply build/bin.NOTICE:187]: Sorry, I didn't see the other comments until after I posted the above, but I think it answers your questions.

> Also in the PR for daffodil some of the items that have no NOTICE in JAR have the license mentioned in bin.LICENSE so could those just be removed from the bin.NOTICE? Some of them also have Apache licenses so should those be removed from the bin.NOTICE file?

I suggested to John that we do not include anything in bin.NOTICE for dependencies that do not have a NOTICE file. I'm not sure if a final decision has been made about that, but that is my preference.

File: build.sbt
Comment (mbeckerle) [inline build.sbt:42]: This change is, FYI, super important because the latest LTS Java version, 17, requires this version of the scala libraries or obscure errors occur. 

And there are good reasons to want Java 17, including that it is noticibly faster.

**Review by mbeckerle — APPROVED** [review]
+1 I have one suggestion due to awkward wording... maybe.

File: build/bin.NOTICE
Comment (stevedlawrence) [inline build/bin.NOTICE:2]: This has already been fixed in main. Might get a merge conflict for this?

File: build/bin.NOTICE
Comment (stevedlawrence) [inline build/bin.NOTICE:82]: This line still exists in a few places. This can be removed in all places except for the very first one at the top.

File: build/bin.NOTICE
Comment (stevedlawrence) [inline build/bin.NOTICE:196]: We need this Http stuff since it is a dependency of Daffodil, but that's weird. Not sure why http stuff is a depdency of Daffodil. I wonder if that's an optional transitive dependency that we can remove in Daffodil 3.4.0?

File: build/bin.NOTICE
Comment (stevedlawrence) [inline build/bin.NOTICE:198]: This is already above. One of those can be removed.

File: build/bin.NOTICE
Comment (stevedlawrence) [inline build/bin.NOTICE:208]: The bin.NOTICE file @tuxji just merged does not list Dumbster or some of these other things from Log4j. Did Log4j change their notice?

@tuxji, it would help if you could take a scan at this. You're most faimilar with all the recent changes.

Also, it's a little hard to just diff since the format has changed. Do we want to standardize all of our license/notice files to the same format?

**Review by stevedlawrence — CHANGES_REQUESTED** [review]

File: build/bin.NOTICE
Comment (mbeckerle) [reply build/bin.NOTICE:208]: Eventually I think a directory of License files, and a directory of Notices files would be better than this squashing of stuff together into a single license and notices file. It's very hard to verify that this includes everything it is supposed to, and does not include things it's not supposed to, as well as not knowing what's verbatim we're including, and what is stuff we're writing.

Comment (stevedlawrence) [reply build/bin.NOTICE:208]: Yeah, we originally just copied from what other projects do, but it should be fine to have a license directory, and is definitely easier to manage. However, I'm not sure you can do the same for NOTICES. I think those have to be in the NOTICE file.

File: build/bin.LICENSE
Comment (tuxji) [reply build/bin.LICENSE:2212]: Actually we can remove the Saxon license because daffodil-debugger does not depend on daffodil-schematron, and therefore does not include Saxon or its transitive dependencies in the daffodil-debugger zip file.  It's important to build the extension and then look inside daffodil-debugger.zip to see which jars daffodil-debugger actually uses instead of simply starting with a copy of Daffodil's bin.LICENSE and bin.NOTICE.  I was surprised how many differences I found between Daffodil's libs and daffodil-debugger's libs once I actually compared the libs.  I don't simply see a superset or subset of one's libs in the other's libs; some dependencies even have different versions.  We only just released Daffodil 3.3.0, so there hasn't been enough time for Daffodil to bump some dependency versions to newer versions yet.  Someone will have to look closely at how daffodil-debugger specifies its dependencies and figure out why certain dependencies have different versions.

P.S.  I know why the versions are different now - when I built the extension, I'd checked out the main branch but the main branch is still using Daffodil 3.2.1, not 3.3.0.

File: build/bin.LICENSE
Comment (tuxji) [reply build/bin.LICENSE:2216]: Saxon is gone now, but I've rearranged the lines mentioning each jar and its license to make it easier to understand which files are covered by which licenses.

File: build/bin.NOTICE
Comment (tuxji) [reply build/bin.NOTICE:82]: I've removed these lines in my edits.

File: build/bin.NOTICE
Comment (tuxji) [reply build/bin.NOTICE:168]: Yes, I removed the dependencies that do not have a NOTICE file in my Daffodil PR's final changes as Steve suggested.  I'll  upload my suggested edits directly to this PR as well.

File: build/bin.NOTICE
Comment (tuxji) [reply build/bin.NOTICE:196]: I removed the http notices because they're not used by daffodil-debugger.  They're in Daffodil's bin.NOTICE only because daffodil-schematron uses Saxon-HE, Saxon-HE uses xml-resolver, and xml-resolver uses these http jars.  You have to build the extension and look at daffodil-debugger's libs as the source of truth before you edit bin.LICENSE or bin.NOTICE.

File: build/bin.NOTICE
Comment (tuxji) [reply build/bin.NOTICE:198]: Done.

File: build/bin.NOTICE
Comment (tuxji) [reply build/bin.NOTICE:208]: Aye, I took a pass at this and updated bin.LICENSE and bin.NOTICE.  I noticed that many jars don't have LICENSE or NOTICE files in them but I didn't take the extra time to visit their source repositories and check for LICENSE/NOTICE files there as well, so I don't promise that my edits are the last word.  I just want to give these files a better foundation for the 1.0.0 release and let future PRs improve them.

**Review by tuxji — COMMENTED** [review]
I've done a pass over both bin.LICENSE and bin.NOTICE to bring them up to date and I will upload my changes to these files using "Edit File" in GitHub's PR user interface so all of you can review my suggested changes.  I made too many changes to simply add comments suggesting what to change in every place.    

In general, I made it easier to see exactly which files are covered by which licenses/notices with better formatting & wording and I sorted all the licenses/notices alphabetically by their jar names to establish a standard ordering in each file.  I'll upload my edits for review next.

**Review by tuxji — APPROVED** [review]
+1

If these bin.LICENSE/bin.NOTICE changes look OK, then please rebase, squash to one commit, and merge.

**Comment by Shanedell** [comment]
@mbeckerle @stevedlawrence Mike and Steve, I think the changes added by @tuxji  look good to me, what do you guys think? Want to make sure to get your input before rebasing, squashing and merging.

**Comment by mbeckerle** [comment]
+1 I'm good with these changes.

**Comment by stevedlawrence** [comment]
+1, looks good

**Pull request merged**
